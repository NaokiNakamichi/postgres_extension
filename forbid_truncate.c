#include "postgres.h"
#include "tcop/utility.h"
#include "utils/builtins.h"

PG_MODULE_MAGIC;


void _PG_init(void);
void _PG_fini(void);
static ProcessUtility_hook_type prev_utility_hook = NULL;
static void forbid_truncate(PlannedStmt *pstmt,
							   const char *queryString,
							   ProcessUtilityContext context,
							   ParamListInfo params,
							   QueryEnvironment *queryEnv,
							   DestReceiver *dest,
							   QueryCompletion *qc);




static void forbid_truncate
(PlannedStmt *pstmt,
						const char *queryString,
						ProcessUtilityContext context,
						ParamListInfo params,
						QueryEnvironment *queryEnv,
						DestReceiver *dest,
						QueryCompletion *qc){
      Node	*parsetree = pstmt->utilityStmt;

       if (nodeTag(parsetree) ==  T_TruncateStmt){
         ereport(ERROR, (errmsg("forbid truncate")));
       }
       else
       standard_ProcessUtility(pstmt, queryString,
								context, params, queryEnv,
								dest, qc);


    }

void _PG_init(void)
{
  /* activate hook module is loaded */
  prev_utility_hook = ProcessUtility_hook;
  ProcessUtility_hook = forbid_truncate;
}

void _PG_fini(void)
{
  ProcessUtility_hook = prev_utility_hook;
}
