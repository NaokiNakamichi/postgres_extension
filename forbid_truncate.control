# forbid_truncate extension
comment = 'forbid truncate'
default_version = '1.0'
module_pathname = '$libdir/forbid_truncate'
relocatable = true
