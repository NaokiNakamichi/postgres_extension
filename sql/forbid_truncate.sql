LOAD 'forbid_truncate';
CREATE TEMP TABLE test (a int, b char(20));

INSERT INTO test VALUES(generate_series(1, 10), 'aaa');

TRUNCATE TABLE test;

SELECT * FROM test;

ALTER TABLE test ADD
	foo VARCHAR(30)
;

ALTER TABLE test ALTER COLUMN
	foo TYPE VARCHAR(20)
;

ALTER TABLE test DROP
	foo
;

ALTER TABLE test
RENAME TO test2;

DROP TABLE test2;
