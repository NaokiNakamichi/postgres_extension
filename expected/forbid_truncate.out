LOAD 'forbid_truncate';
CREATE TEMP TABLE test (a int, b char(20));
INSERT INTO test VALUES(generate_series(1, 10), 'aaa');
TRUNCATE TABLE test;
ERROR:  forbid truncate
SELECT * FROM test;
 a  |          b           
----+----------------------
  1 | aaa                 
  2 | aaa                 
  3 | aaa                 
  4 | aaa                 
  5 | aaa                 
  6 | aaa                 
  7 | aaa                 
  8 | aaa                 
  9 | aaa                 
 10 | aaa                 
(10 rows)

ALTER TABLE test ADD
	foo VARCHAR(30)
;
ALTER TABLE test ALTER COLUMN
	foo TYPE VARCHAR(20)
;
ALTER TABLE test DROP
	foo
;
ALTER TABLE test
RENAME TO test2;
DROP TABLE test2;
